lcli/src/etl/block_efficiency.rs

```diff
@@ -0,0 +1,317 @@
+        // as determined by the committee size/number.
+        // This is unlikely to happen, but not impossible.
+        let max_possible_attesations =
+            committee_data.1.validators_per_committee * committee_data.1.number_of_committees;
+
+        // Get number of active validators.
+        let active_validators =
+            get_validator_set_len::<T>(&node, epoch.start_slot(T::slots_per_epoch())).await?;
+
+        let start_slot = epoch.start_slot(T::slots_per_epoch());
+        for step in 0..32 {
+            let slot = start_slot + step;
+
+            // Get all included attestations.
+            let (mut attestations_in_block, proposer) =
+                match get_block_attestations_set::<T>(&node, slot).await {
+                    Ok(output) => (output.0, Some(output.1)),
+                    Err(_) => (HashMap::new(), None),
+                };
+
+            // Insert block proposer into proposer_map.
+            if let Some(proposer_info) = proposer {
+                proposer_map.insert(slot, proposer_info.clone());
+            }
+
+            // Remove duplicate attestations.
+            attestations_in_block.retain(|x, _| included_attestations_set.get(x).is_none());
+
+            // Add them to the set.
+            included_attestations_set.extend(attestations_in_block.clone());
+
+            // Don't write data from the initialization epoch.
+            if epoch != initialization_epoch {
+                let included = attestations_in_block.len();
+
+                let available = if max_possible_attesations < available_attestations_set.len() {
+                    max_possible_attesations
+                } else {
+                    available_attestations_set.len()
+                };
+
+                // Get proposer information.
+                let proposer = proposer_map.get(&slot).cloned();
+
+                // Store slot data.
+                epoch_data.push((slot, proposer, available, included));
+            }
+
+            // Included attestations are no longer available.
+            for new_attestation in &attestations_in_block {
+                available_attestations_set.remove(new_attestation.0);
+            }
+
+            // Get all available attestations.
+            for committee in &committee_data.0 {
+                if committee.slot == slot {
+                    for position in 0..committee.validators.len() {
+                        let unique_attestation = UniqueAttestation {
+                            slot: committee.slot,
+                            committee_index: committee.index,
+                            committee_position: position,
+                        };
+                        available_attestations_set.insert(unique_attestation.clone());
+                    }
+                }
+            }
+
+            // Remove expired available attestations.
+            available_attestations_set.retain(|x| x.slot >= (slot - 32));
```

Comment:
`slot - 32` seems at risk of an underflow. When running without `debug_assertions` (i.e., in `release`), Rust will silently underflow.

Reply:
This should only affect epoch 0 (so only when running with `--start-epoch 1`), so a `saturating_sub` here should be fine, although in my testing it didn't affect any results :thinking: